docs/design-documents/drivers/serial/serial.md

```diff
@@ -0,0 +1,321 @@
+
+    /** Write a char to the serial port
+     *
+     * @param c The char to write
+     *
+     * @returns The written char or -1 if an error occurred
+     */
+    int putc(int c);
+
+    /** Read a char from the serial port
+     *
+     * @returns The char read from the serial port
+     */
+    int getc();
+
+    /** Write a string to the serial port
+     *
+     * @param str The string to write
+     *
+     * @returns 0 if the write succeeds, EOF for error
+     */
+    int puts(const char *str);
+
+    virtual ssize_t write(const void *buffer, size_t size);
+    virtual ssize_t read(void *buffer, size_t size);
+    virtual off_t seek(off_t offset, int whence = SEEK_SET);
+    virtual off_t size();
+    virtual int isatty();
+    virtual int close();
+    virtual short poll(short events) const;
+
+#if !(DOXYGEN_ONLY)
+protected:
+
+    /* Acquire exclusive access to this serial port
+     */
+    virtual void lock(void);
+
+    /* Release exclusive access to this serial port
+     */
+    virtual void unlock(void);
+#endif
+};
+```
+
+###  Detailed design : MinimalSerial
+
+The `MinimalSerial` class is a minimal abstract base class that provides basics I/O functionality. All other serial classes derive from it.
+
+**API description**
+
+This is the proposed API for the `MinimalSerial` class. Only two methods are provided to output `_base_putc()` and receive `_base_getc()` characters on the UART.
+
+```C
+class MinimalSerial : private NonCopyable<MinimalSerial> {
+protected:
+    MinimalSerial(PinName tx, PinName rx, int baud);
+    virtual ~MinimalSerial();
+    int _base_getc();
+    int _base_putc(int c);
+};
+```
+
+###  Detailed design : MinimalConsole
+
+`MinimalConsole` is an internal concrete class that derives from `MinimalSerial` and provides basic console functionality that can be used by the retarget code for resources constrained targets, see [Simplication of retarget code](#simplification-of-retarget-code).
+
+**API description**
+
+```C
+class MinimalConsole : public MinimalSerial, private NonCopyable<MinimalConsole> {
+public:
+    MinimalConsole(PinName tx, PinName rx, int baud);
+    virtual ~MinimalConsole();
+    int getc();
+    int putc(int c);
+};
+```
+
+The constructor provides arguments to create a serial port connected to the specified transmit and receive pins and set to the specified baud rate. If a target has serial support, by default a serial port is used for the console and the pins and settings for the port selection come from target header files and JSON settings. 
+
+# Usage scenarios and examples
+
+### Scenario 1 `Printing to stdout`
+
+The below example shows pseudocode for how to print `Hello` to stdout using the new API. Printf cannot be called directly; instead the application first calls _fdopen_ to get a _FILE*_ to use the printf function.
+
+```C
+BufferedSerial serial(RX, TX, 115200);
+FILE *out = fdopen(&serial, "w");
+fprintf(out, "Hello");
+```
```

Comment:
That's not printing to `stdout`. Printing to `stdout` would be

    printf("Hello");

or

    fprintf(stdout, "Hello");

Users should not be messing with the serial port used by the console - that's bad practice, and exactly one of the things in our examples we should be discouraging.

Previously people did that because it was the only way of setting the console baud rate, but that's no longer true - can be configured via JSON.

This example would be valid for printing to a different serial port.

> Printf cannot be called directly

It absolutely can. I think you mean to say you can't call `serial.printf`?

Reply:
Thanks for both corrections. On the last point, I did mean to say you can't call `serial.printf`. 
I will add two scenarios: one to print to stdout and one to print to a serial port.